util/resourcetoplogy/topology.go

```diff
@@ -0,0 +1,391 @@
+			}
+			for _, item := range items {
+				resources = append(resources, k8s.ResourceIdentifier{
+					Group:     selector.Group,
+					Resource:  selector.Resource,
+					Name:      item.GetName(),
+					Namespace: item.GetNamespace(),
+				})
+			}
+		}
+	}
+	return resources, nil
+}
+
+func (r *engine) handleBuiltInRules(ctx context.Context, typ string, v cue.Value, resource k8s.ResourceIdentifier) ([]k8s.ResourceIdentifier, error) {
+	switch strings.ToLower(typ) {
+	case "service":
+		return r.handleBuiltInRulesForService(ctx, v, resource)
+	default:
+		return nil, fmt.Errorf("unsupported built-in rule %s", typ)
+	}
+}
+
+func (r *engine) getGroupResourceFromSubs(sub SubResource, group, resource string) []k8s.ResourceIdentifier {
+	result := make([]k8s.ResourceIdentifier, 0)
+	if sub.ResourceIdentifier.Group == group && sub.ResourceIdentifier.Resource == resource {
+		result = append(result, sub.ResourceIdentifier)
+	}
+	for _, child := range sub.Children {
+		result = append(result, r.getGroupResourceFromSubs(child, group, resource)...)
+	}
+	return result
+}
+
+func (r *engine) handleBuiltInRulesForService(ctx context.Context, v cue.Value, resource k8s.ResourceIdentifier) ([]k8s.ResourceIdentifier, error) {
+	subs, err := r.getSubResources(ctx, v, resource)
+	if err != nil {
+		return nil, err
+	}
+	pods := make([]k8s.ResourceIdentifier, 0)
+	for _, sub := range subs {
+		pods = append(pods, r.getGroupResourceFromSubs(sub, "", "pod")...)
+	}
+	// get service endpoints and compare with pods
+	es := &discoveryv1.EndpointSliceList{}
+	if err = singleton.KubeClient.Get().List(ctx, es, client.InNamespace(resource.Namespace)); err != nil {
+		return nil, err
+	}
+	service := []k8s.ResourceIdentifier{}
+	for _, e := range es.Items {
+		for _, s := range e.Endpoints {
+			if slices.Contains(pods, k8s.ResourceIdentifier{
+				Name:      s.TargetRef.Name,
+				Namespace: s.TargetRef.Namespace,
+				Group:     "",
+				Resource:  strings.ToLower(s.TargetRef.Kind),
+			}) {
+				service = append(service, k8s.ResourceIdentifier{
+					Group:     "",
+					Resource:  "service",
+					Name:      e.OwnerReferences[0].Name,
+					Namespace: resource.Namespace,
+				})
+			}
+		}
+	}
+	return service, nil
+}
+
+func listResources(ctx context.Context, selector *ResourceSelector, relation k8s.ResourceIdentifier) ([]unstructured.Unstructured, error) {
+	cli := singleton.KubeClient.Get()
+	resource := k8s.ResourceIdentifier{
+		Group:    selector.Group,
+		Resource: selector.Resource,
+	}
+	listOpts := make([]client.ListOption, 0)
+	var annos map[string]string
+	var owner bool
+	switch selector.SelectorKey {
+	case namespaceSelectorKey:
+		if ns, err := selector.SelectorValue.String(); err == nil {
+			listOpts = append(listOpts, client.InNamespace(ns))
+		}
+	case labelsSelectorKey:
+		labels := make(map[string]string)
+		if err := selector.SelectorValue.Decode(&labels); err == nil {
+			listOpts = append(listOpts, client.MatchingLabels(labels))
+		}
```

Comment:
Can we combine two selector ways eg: namespace and lablesSelector?

Reply:
I think we should keep selector simple, if the users want some complicated selectors, we can support listOptions in the future as a kind of selector.

Reply:
Currently, we need to use the namespace and labelSelector to list the pods of the ReplicaSet.